cupdlp/cupdlp_solver.c

```diff
@@ -68,6 +80,39 @@ void PDHG_Compute_Dual_Feasibility(CUPDLPwork *work, double *dualResidual,
   //    *dDualObj = Dotprod_Neumaier(problem->rhs, y, lp->nRows);
   cupdlp_dot(work, lp->nRows, y, problem->rhs, dDualObj);
 
+#if !(CUPDLP_CPU) && USE_KERNELS
+  dualResidual = work->buffer2;
+
+  cupdlp_float alpha;
+  cupdlp_dual_feasibility_kernel_1_cuda(dualResidual, aty,
+                                        problem->cost, lp->nCols);
+
+  cupdlp_dual_feasibility_kernel_2_cuda(dSlackPos, dualResidual,
+                                        problem->hasLower, lp->nCols);
```

Comment:
The overhead from starting kernels is very low, so "whenever possible" is perhaps overdoing it :). I do understand what you mean though; for these kernels we could avoid loading `dualResidual` one extra time from memory if fusing.
But these kernels (called only from `PDHL_Compute_Dual_Feasibility`) are not called from the innermost loop, and `dualResidual` is read a number of times (I count six) in the function `PDHL_Compute_Dual_Feasibility`. My guess is therefore that it would not make a noticeable difference to fuse the two kernels. Of course, fusing the kernels would not make the code much more complicated, so maybe we want to do it even if no performance is won. The reason I did not is that the fused kernel would have two "output vectors", which I think adds a little conceptual messiness, and that I prefer short kernels if possible.
If you like I can run some tests to see if there is something to gain here.